File: lualib/xodel/dotenv.lua
local gmatch = ngx.re.gmatch
local match = ngx.re.match


-- https://github.com/motdotla/dotenv/blob/master/lib/main.js
local LINE = [[(?:^|^)\s*(?:export\s+)?([\w.-]+)(?:\s*=\s*?|:\s+?)(\s*'(?:\\'|[^'])*'|\s*"(?:\\"|[^"])*"|\s*`(?:\\`|[^`])*`|[^#\r\n]+)?\s*(?:#.*)?(?:$|$)]]

local function dict(a, b)
  local t = {}
  for key, value in pairs(a) do
    t[key] = value
  end
  for key, value in pairs(b) do
    t[key] = value
  end
  return t
end

local function split(s, sep)
  local res = {}
  sep = sep or " "
  local i = 1
  local a, b
  while true do
    a, b = s:find(sep, i, true)
    if a then
      local e = s:sub(i, a - 1)
      i = b + 1
      res[#res + 1] = e
    else
      res[#res + 1] = s:sub(i)
      return res
    end
  end
end

local function parse(src)
  local obj = {}
  local it = assert(gmatch(src, LINE, "msui"))
  while true do
    local m, err = it()
    if err then
      return nil, err
    end
    if not m then
      break
    end
    local key = m[1]
    local value = m[2] or ""
    value = (value:gsub("^%s*(.-)%s*$", "%1"))
    local maybeQuote = value:sub(1, 1);
    -- remove surrounding quotes
    value = (value:gsub([[^(['"`])([%s%S]*)%1$]], "%2"))
    if maybeQuote == '"' then
      value = value:gsub([[\n]], "\n")
      value = value:gsub([[\r]], "\r")
    end
    obj[key] = value;
  end
  return obj;
end

---@param env_value string
---@param env table
---@return string
local function expand_key(env_value, env)
  local index = 0
  for m in gmatch(env_value, [[(.?\${*[\w]*(?::-[\w/]*)?}*)]], 'msui') do
    index = index + 1
    local parts = assert(match(m[1], [[(.?)\${*([\w]*(?::-[\w/]*)?)?}*]]))
    local prefix = parts[1]
    local value, replacePart;
    if prefix == '\\' then
      ---@type string
      replacePart = parts[1];
      value = replacePart:sub(2)
    else
      local keyParts = split(parts[2], ':-')
      local key = keyParts[1]
      replacePart = parts[0]:sub(#prefix + 1);
      local ek = os.getenv(key)
      if ek ~= nil then
        value = ek
      else
        value = env[key] or keyParts[2] or ""
      end
      -- if #keyParts>1 and value then
      --   local replaceNested = ""
      -- end
      value = expand_key(value, env)
    end
    env_value = assert(ngx.re.gsub(env_value, '\\' .. replacePart, value))
  end
  return env_value
end

local function expand(env)
  for key, value in pairs(env) do
    env[key] = expand_key(value, env)
  end
  return env
end

local function make(opts)
  opts = opts or {}
  local path
  local env = {}
  if opts.path == nil then
    path = '.env'
  elseif type(opts.path) == 'string' then
    path = opts.path
  elseif type(opts.path) == 'table' then
    for _, p in ipairs(opts.path) do
      for key, value in pairs(make(dict(opts, { path = p }))) do
        env[key] = value
      end
    end
  else
    error("invald path type:" .. type(opts.type))
  end
  if path then
    local content = assert(io.open(path, "r")):read("*a")
    env = dict(env, assert(parse(content)))
  end
  return env
end

-- local function say(...)
--   for index, value in ipairs({ ... }) do
--     ngx.print(require('xodel.utils').repr(value))
--   end
--   ngx.say('---------------------------------')
-- end

local function __call(t, opts)
  return expand(make(opts))
end

local dotenv = setmetatable({ parse = parse, make = make, expand = expand }, { __call = __call })

dotenv.__index = dotenv

return dotenv

File: bin/dotenv.js
const fs = require("fs");
const path = require("path");
const os = require("os");
const packageJson = require("../package.json");

const version = packageJson.version;

const LINE =
  /(?:^|^)\s*(?:export\s+)?([\w.-]+)(?:\s*=\s*?|:\s+?)(\s*'(?:\\'|[^'])*'|\s*"(?:\\"|[^"])*"|\s*`(?:\\`|[^`])*`|[^#\r\n]+)?\s*(?:#.*)?(?:$|$)/gm;

// Parser src into an Object
function parse(src) {
  const obj = {};

  // Convert buffer to string
  let lines = src.toString();

  // Convert line breaks to same format
  lines = lines.replace(/\r\n?/gm, "\n");

  let match;
  while ((match = LINE.exec(lines)) != null) {
    const key = match[1];

    // Default undefined or null to empty string
    let value = match[2] || "";

    // Remove whitespace
    value = value.trim();

    // Check if double quoted
    const maybeQuote = value[0];

    // Remove surrounding quotes
    value = value.replace(/^(['"`])([\s\S]*)\1$/gm, "$2");

    // Expand newlines if double quoted
    if (maybeQuote === '"') {
      value = value.replace(/\\n/g, "\n");
      value = value.replace(/\\r/g, "\r");
    }

    // Add to object
    obj[key] = value;
  }

  return obj;
}

function _log(message) {
  console.log(`[dotenv@${version}][DEBUG] ${message}`);
}

function _resolveHome(envPath) {
  return envPath[0] === "~"
    ? path.join(os.homedir(), envPath.slice(1))
    : envPath;
}

// Populates process.env from .env file
function config(options) {
  let dotenvPath = path.resolve(process.cwd(), ".env");
  let encoding = "utf8";
  const debug = Boolean(options && options.debug);
  const override = Boolean(options && options.override);

  if (options) {
    if (options.path != null) {
      dotenvPath = _resolveHome(options.path);
    }
    if (options.encoding != null) {
      encoding = options.encoding;
    }
  }

  try {
    // Specifying an encoding returns a string instead of a buffer
    const parsed = DotenvModule.parse(
      fs.readFileSync(dotenvPath, { encoding })
    );

    Object.keys(parsed).forEach(function (key) {
      if (!Object.prototype.hasOwnProperty.call(process.env, key)) {
        process.env[key] = parsed[key];
      } else {
        if (override === true) {
          process.env[key] = parsed[key];
        }

        if (debug) {
          if (override === true) {
            _log(
              `"${key}" is already defined in \`process.env\` and WAS overwritten`
            );
          } else {
            _log(
              `"${key}" is already defined in \`process.env\` and was NOT overwritten`
            );
          }
        }
      }
    });

    return expand({ parsed });
  } catch (e) {
    if (debug) {
      _log(`Failed to load ${dotenvPath} ${e.message}`);
    }

    return { error: e };
  }
}

const DotenvModule = {
  config,
  parse,
};

function _interpolate(envValue, environment, config) {
  const matches = envValue.match(/(.?\${*[\w]*(?::-[\w/]*)?}*)/g) || [];
  return matches.reduce(function (newEnv, match, index) {
    const parts = /(.?)\${*([\w]*(?::-[\w/]*)?)?}*/g.exec(match);
    if (!parts || parts.length === 0) {
      return newEnv;
    }
    const prefix = parts[1];

    let value, replacePart;

    if (prefix === "\\") {
      replacePart = parts[0];
      value = replacePart.replace("\\$", "$");
    } else {
      const keyParts = parts[2].split(":-");
      const key = keyParts[0];
      replacePart = parts[0].substring(prefix.length);
      // process.env value 'wins' over .env file's value
      value = Object.prototype.hasOwnProperty.call(environment, key)
        ? environment[key]
        : config.parsed[key] || keyParts[1] || "";

      // If the value is found, remove nested expansions.
      if (keyParts.length > 1 && value) {
        const replaceNested = matches[index + 1];
        matches[index + 1] = "";

        newEnv = newEnv.replace(replaceNested, "");
      }
      // Resolve recursive interpolations
      value = _interpolate(value, environment, config);
    }

    return newEnv.replaceAll(replacePart, value);
  }, envValue);
}

function expand(config) {
  // if ignoring process.env, use a blank object
  const environment = config.ignoreProcessEnv ? {} : process.env;

  for (const configKey in config.parsed) {
    const value = Object.prototype.hasOwnProperty.call(environment, configKey)
      ? environment[configKey]
      : config.parsed[configKey];

    config.parsed[configKey] = _interpolate(value, environment, config);
  }

  for (const processKey in config.parsed) {
    environment[processKey] = config.parsed[processKey];
  }

  return config;
}

console.log(config());
module.exports.expand = expand;
module.exports.config = DotenvModule.config;
module.exports.parse = DotenvModule.parse;
module.exports = DotenvModule;
